### PR TITLE
pod-overhead: add documentation

### PR DESCRIPTION
**Pull request by egernst**
Initial documentation for PodOverhead feature in Kubernetes 1.16 release

### CODE REVIEW

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+2. Add a pod Overhead to the defined RuntimeClass
+
+#### 1. Configure basic RuntimeClass
+
+
+#### 2. Add pod Overhead to RuntimeClass definition
+
+
+{{< note >}}
+It is recommended that RuntimeClass write operations (create/update/patch/delete) be
```
Comment (sftim) [inline content/en/docs/concepts/containers/pod-overhead.md]: How about this wording, or something like it:
```markdown
Users who can write to RuntimeClass resources are able to have cluster-wide impact on
workload performance. You can limit access to this ability using Kubernetes access controls.
See …
```
(aim here is to avoid writing in the passive voice&mdash;“It is recommended”).

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+restricted to the cluster administrator. This is typically the default. See [Authorization
+Overview](/docs/reference/access-authn-authz/authorization/) for more details.
+{{< /note >}}
+
```
Comment (sftim) [inline content/en/docs/concepts/containers/pod-overhead.md]:
```suggestion
{{% /capture %}}
```
(for tidiness)

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+This page describes pod Overhead, as defined in RuntimeClass and pod.Spec, and mechanisms for applying this feature.
+
+{{% /capture %}}
+
+
+{{% capture body %}}
+
+## Pod Overhead
+
+Pod Overhead is a feature for applying an overhead to a pods definition at admission time depending on the RuntimeClass being utilized.
```
Comment (sftim) [inline content/en/docs/concepts/containers/pod-overhead.md]: This line feels like a great fit for the overview section.

Also, how about adding a hyperlink to “admission”:
```suggestion
Pod Overhead is a feature for applying an overhead to a pods definition at [admission](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks) time, depending on the RuntimeClass being utilized.
```

If the link doesn't feel right, maybe I can find or tweak another page that explains _admission_ as a Kubernetes concept.

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+### Set Up
+
+Ensure the PodOverhead feature gate is enabled (it is by default). See [Feature
+Gates](/docs/reference/command-line-tools-reference/feature-gates/) for an explanation of enabling
+feature gates. The `PodOverhead` feature gate must be enabled on scheduler, apiservers _and_ kubelets.
+
+1. Configure basic RuntimeClass
+2. Add a pod Overhead to the defined RuntimeClass
+
+#### 1. Configure basic RuntimeClass
```
Comment (sftim) [inline content/en/docs/concepts/containers/pod-overhead.md]: I'd consider stopping the Concept page at this point, and hyperlinking to a new, separate Task page about using PodOverhead.

(it's not a problem for Concept pages to be short; my gut feeling is that that's actually better)

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+- dchen1107
+- egernst
+title: Pod Overhead
+content_template: templates/concept
+weight: 20
+---
+
+{{% capture overview %}}
+
+{{< feature-state for_k8s_version="v1.16" state="alpha" >}}
```
[the remark this answers is not in the document]
Comment (tallclair) [reply content/en/docs/concepts/containers/pod-overhead.md]: That's correct. I think it's good to call that out and link to the feature gates doc from here.

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+This page describes pod Overhead, as defined in RuntimeClass and pod.Spec, and mechanisms for applying this feature.
```
Comment (tallclair) [inline content/en/docs/concepts/containers/pod-overhead.md]: nit: inconcistent capitalization. Stick with either `Pod Overhead` or `pod overhead`

File: content/en/docs/concepts/containers/pod-overhead.md (hunk 3 above)
Comment (tallclair) [inline content/en/docs/concepts/containers/pod-overhead.md]: I'd start by explaining what overhead is, and why it's important. Maybe:
```suggestion
Pod Overhead is a feature for accounting for the resources consumed by the pod infrastructure on top of the container requests & limits. The pod's overhead is set at admission time according to the overhead associated with the pod's [RuntimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/).
```

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+{{% capture body %}}
+
+## Pod Overhead
+
+Pod Overhead is a feature for applying an overhead to a pods definition at admission time depending on the RuntimeClass being utilized.
+
+### Set Up
+
+Ensure the PodOverhead feature gate is enabled (it is by default). See [Feature
```
Comment (tallclair) [inline content/en/docs/concepts/containers/pod-overhead.md]: Alpha features are off by default.
```suggestion
Ensure the PodOverhead feature gate is enabled (it is disabled by default). See [Feature
```

File: content/en/docs/concepts/containers/pod-overhead.md (hunk 2 above)
Comment (tallclair) [inline content/en/docs/concepts/containers/pod-overhead.md]: Can we add a link to the KEP to learn more at the end of the page? Is that something that is done? (@sftim)

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+Gates](/docs/reference/command-line-tools-reference/feature-gates/) for an explanation of enabling
+feature gates. The `PodOverhead` feature gate must be enabled on scheduler, apiservers _and_ kubelets.
+
+1. Configure basic RuntimeClass
+2. Add a pod Overhead to the defined RuntimeClass
+
+#### 1. Configure basic RuntimeClass
+
+
+#### 2. Add pod Overhead to RuntimeClass definition
```
Comment (tallclair) [inline content/en/docs/concepts/containers/pod-overhead.md]: Can you also add a section to the RuntimeClass page? 1-2 sentences plus a link to this document should be sufficient.

File: content/en/docs/concepts/containers/pod-overhead.md (hunk 2 above)
Comment (sftim) [reply content/en/docs/concepts/containers/pod-overhead.md]: Linking to the KEP is relevant, and a `whatsnext` section is a great place to put relevant links.

File: content/en/docs/concepts/containers/pod-overhead.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
+---
```
Comment (sftim) [inline content/en/docs/concepts/containers/pod-overhead.md]: I would expect to see this filed inside https://kubernetes.io/docs/concepts/workloads/pods/

File: content/en/docs/concepts/containers/pod-overhead.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/concepts/containers/pod-overhead.md]: Use italics for the first mention of “Pod Overhead”. Here's a full example that frames _Pod Overhead_ as a general concept with an implementation in Kubernetes.
```markdown
When you run a Pod on a Node, the Pod itself takes an amount of system resources. These
resource are additional to the resources needed to run the container(s) inside the Pod.
_Pod Overhead_ is a way of accounting for the extra resources consumed by the Pod itself.

Kubernetes can track the overhead for Pods, and account for these resources. In Kubernetes,
a Pod's overhead is set at admission time, based on the overhead settings for that
Pod's [RuntimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/).
```

**Comment by sftim** [comment]
Because this PR currently has placeholder content, I will
/retitle [WIP] pod-overhead: initial placeholder documentation 

and I hope that's OK.

(If the idea was to merge this but then either revert or submit another PR, that might be fine - I'm not really clear on how the SIG Docs workflow ties in with the regular Kubernetes release process)

**Comment by VineethReddy02** [comment]
Hey, @egernst @sftim we have a deadline for this PR to be in ready to be merged state by 7th september i.e tomorrow. Please make sure we are following the deadline.

**Comment by egernst** [comment]
Thanks for all of the feedback.  I'll have a PR later this morning!

**Comment by sftim** [comment]
I probably won't be able to give any more feedback on this PR until after the v1.16 release.

File: content/en/docs/concepts/containers/pod-overhead.md (hunk 9 above)
Comment (tallclair) [reply content/en/docs/concepts/containers/pod-overhead.md]: Or maybe https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/ makes more sense?

Comment (egernst) [reply content/en/docs/concepts/containers/pod-overhead.md]: agreed.  changed accordingly. PTAL

**Comment by egernst** [comment]
Updated based on feedback.  PTAL

**Comment by simplytunde** [comment]
Previews

- https://5d72ac0b04faa000083164c5--kubernetes-io-vnext-staging.netlify.com/docs/concepts/containers/runtime-class/
- https://5d72ac0b04faa000083164c5--kubernetes-io-vnext-staging.netlify.com/docs/concepts/configuration/pod-overhead/

File: content/en/docs/concepts/containers/runtime-class.md [hunk 10]
```diff
@@ -171,6 +171,18 @@ Nodes](/docs/concepts/configuration/assign-pod-node/).
 
 [admission controller]: /docs/reference/access-authn-authz/admission-controllers/
 
+### Pod Overhead
+
+{{< feature-state for_k8s_version="v1.16" state="alpha" >}}
+
+As of Kubernetes v1.16, RuntimeClass includes support for specifying overhead associated with
+running a pod, as part of the [`PodOverhead`](/docs/concepts/configuration/pod-overhead.md) feature.
+To use `PodOverhead`, you must have the PodOverhead feature gate enabled (it is off by default).
```
Comment (sftim) [inline content/en/docs/concepts/containers/runtime-class.md]: nit:
```suggestion
To use `PodOverhead`, you must have the PodOverhead [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) enabled (it is off by default).
```

File: content/en/docs/concepts/configuration/pod-overhead.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+Users who can write to RuntimeClass resources are able to have cluster-wide impact on
+workload performance. You can limit access to this ability using Kubernetes access controls.
+See [Authorization Overview](/docs/reference/access-authn-authz/authorization/) for more details.
+{{< /note >}}
+
+{{% /capture %}}
+
+{{% capture whatsnext %}}
+
+* [RuntimeClass](/docs/concepts/containers/runtime-class.md)
```
Comment (sftim) [inline content/en/docs/concepts/configuration/pod-overhead.md]: This is a broken hyperlink
(also the next line)

File: content/en/docs/concepts/configuration/pod-overhead.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+- in {{< glossary_tooltip text="kube-apiserver" term_id="kube-apiserver" >}}
+- in the {{< glossary_tooltip text="kubelet" term_id="kubelet" >}} on each Node
+- in any custom API servers that use feature gates
+
+{{< note >}}
+Users who can write to RuntimeClass resources are able to have cluster-wide impact on
+workload performance. You can limit access to this ability using Kubernetes access controls.
+See [Authorization Overview](/docs/reference/access-authn-authz/authorization/) for more details.
+{{< /note >}}
+
```
Comment (sftim) [inline content/en/docs/concepts/configuration/pod-overhead.md]: Does the scheduler make different decisions based on the PodOverhead?

What kind of impact does this overhead have, if I configure it?

Comment (egernst) [reply content/en/docs/concepts/configuration/pod-overhead.md]: Indeed, it does impact scheduling, as well as some kubelet behavior.  I added a short blurb on this, and it is discussed in detail in the design doc (KEP).  I plan to write a more thorough overview of the feature and example usage in a blog post, though I'm not sure if it makes sense here?

File: content/en/docs/concepts/configuration/pod-overhead.md (hunk 11 above)
Comment (egernst) [reply content/en/docs/concepts/configuration/pod-overhead.md]: Fixed, though the second one seemed fine?

**Comment by egernst** [comment]
> I also wonder if Pod Overhead as a concept belongs in Scheduling?

I think current location makes more sense. While it does impact scheduling (overhead is added to the total container requests), it also is considered in other areas, like Kubelet.

File: content/en/docs/concepts/configuration/pod-overhead.md (hunk 12 above)
Comment (sftim) [reply content/en/docs/concepts/configuration/pod-overhead.md]: That makes sense.

_Optional tweak_
How about adding a link from https://kubernetes.io/docs/concepts/scheduling/kube-scheduler/#what-s-next to the new page on Pod Overhead?

Comment (egernst) [reply content/en/docs/concepts/configuration/pod-overhead.md]: Sure. done.

File: content/en/docs/concepts/containers/runtime-class.md [hunk 13]
```diff
@@ -202,5 +216,7 @@ RuntimeClass feature to the beta version:
 
 - [RuntimeClass Design](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class.md)
 - [RuntimeClass Scheduling Design](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class-scheduling.md)
+- [PodOverhead Concept](/docs/concepts/configuration/pod-overhead.md)
```
Comment (sftim) [inline content/en/docs/concepts/containers/runtime-class.md]: One thing: there's no Kubernetes resource called PodOverhead (I can't `kubectl get PodOverhead`). For that reason, the text of the hyperlinks would work better like:
```suggestion
 - Read about the [Pod overhead](/docs/concepts/configuration/pod-overhead/) concept
```
(this also fixes a broken hyperlink)

**Comment by simplytunde** [comment]
@egernst This PR deadline for merging is Sep 9th.

File: content/en/docs/concepts/containers/runtime-class.md (hunk 13 above)
Comment (egernst) [reply content/en/docs/concepts/containers/runtime-class.md]: fixed.

**Comment by simplytunde** [comment]
@egernst  Once the list format and link issue are addressed, it lgtm.

**Comment by egernst** [comment]
@simplytunde can clarify what you’re seeing? I think all issues identified have been addressed?

**Comment by simplytunde** [comment]
@egernst If you look at the preview https://5d73e43ace4f600008594ce3--kubernetes-io-vnext-staging.netlify.com/docs/concepts/configuration/pod-overhead/#set-up, it is not showing up as a list. You need to leave a blank line before each list for it to be properly formatted. https://kubernetes.io/docs/contribute/style/style-guide/#lists. Let me know if I am mistaken and a list is not intended.

**Comment by egernst** [comment]
OK, repushed, sorry - didn't know.

**Comment by simplytunde** [comment]
Thank you so much @egernst. You have been awesome!
/approve
